godot-core/src/meta/param_tuple/impls.rs

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use crate::builtin::Variant;
+use crate::meta::error::{CallError, ConvertError};
+use crate::meta::{signature, CallContext};
+use crate::meta::{
+    FromGodot, GodotConvert, GodotFfiVariant, GodotType, InParamTuple, OutParamTuple, ParamTuple,
+    ToGodot,
+};
+use godot_ffi as sys;
+use std::fmt;
+use sys::GodotFfi;
+
+macro_rules! unsafe_impl_param_tuple {
+    ($Len:literal; $(($p:ident, $n:tt): $P:ident),+) => {
+        impl<$($P),+> ParamTuple for ($($P,)+) where $($P: GodotConvert + fmt::Debug),+ {
+            const LEN: usize = $Len;
```

Comment:
you can count the number of `$P`s with something like this and don't have to manually maintain the count:

```rust
    (count_items; $PP:ident, $($P:ident),+) => {
        1 + unsafe_impl_param_tuple!(count_items; $($P),+)
    };

    (count_items; $P:ident) => {
        1
    };
```
and:
```rust
const LEN: usize = unsafe_impl_param_tuple!(count_items; $($P),+);
```